amqprs/src/api/connection.rs

```diff
@@ -365,6 +416,109 @@ impl OpenConnectionArguments {
     }
 }
 
+#[cfg(feature = "urispec")]
+impl TryFrom<&str> for OpenConnectionArguments {
+    type Error = Error;
+
+    /// Create a new OpenConnectionArguments from a URI &str. Mostly follows the [AMQP URI spec](https://www.rabbitmq.com/uri-spec.html). See below for exceptions.
+    ///
+    /// If the URI is invalid, a UriError error is returned.
+    ///
+    /// If no port is specified, the default port of 5672 is used (as per the spec).
+    /// If no host is speecified or is zero-length, a UriError error is returned. Note that this is different from the spec, which allows for EmptyHost. This is because the host is used to create a TCP connection, and an empty host is invalid.
+    ///
+    fn try_from(uri: &str) -> Result<Self> {
+        let pu = URIReference::try_from(uri)?;
+
+        // Check scheme
+        if pu.scheme().is_none()
+            || (pu.scheme().unwrap().as_str() != "amqp" && pu.scheme().unwrap().as_str() != "amqps")
+        {
+            return Err(Error::UriError(String::from("Invalid URI scheme")));
+        }
+
+
+        // Set the default port depending on the scheme. The unwrap should be safe due to the checks above. 
+        // Will panic if any invalid scheme is not rejected before this point
+        let default_port: &u16 = match pu.scheme().unwrap().as_str() {
+            "amqp" => &DEFAULT_AMQP_PORT,
+            "amqps" => &DEFAULT_AMQPS_PORT,
+            _ => panic!("Error occurred while setting default port based on amq scheme. This should never happen.")
+            
+        };
+
+        // Check authority
+        let pu_authority_result = pu.authority();
+
+        if pu_authority_result.is_none() {
+            return Err(Error::UriError(String::from("Invalid URI authority")));
+        }
+
+        let pu_authority = pu_authority_result.unwrap();
+
+        // Check username
+        let mut pu_authority_username: Option<&str> = None;
+        if let Some(pu_a_u) = pu_authority.username() {
+            pu_authority_username = Some(pu_a_u.as_str());
+        }
+
+        // Check password
+        let mut pu_authority_password: Option<&str> = None;
+        if let Some(pu_a_p) = pu_authority.password() {
+            pu_authority_password = Some(pu_a_p.as_str());
+        }
+
+        // Check virtual host
+        let pu_path = pu.path().to_string();
+
+        // Check query
+        let pu_query = pu.query();
+
+        // Apply authority
+        let mut args: &mut OpenConnectionArguments = &mut OpenConnectionArguments::new(
+            pu_authority.host().to_string().as_str(),
+            pu_authority.port().unwrap_or(*default_port),
+            pu_authority_username.unwrap_or(""),
+            pu_authority_password.unwrap_or(""),
+        );
```

Comment:
to keep default username and password consistent, I think above codes can be replaced with below
```
       let pu_authority_username = pu_authority.username().unwrap_or_else(|| "guest");
       let pu_authority_password = pu_authority.password().unwrap_or_else(|| "guest");

        // Apply authority
        let mut args: &mut OpenConnectionArguments = &mut OpenConnectionArguments::new(
            pu_authority.host().to_string().as_str(),
            pu_authority.port().unwrap_or_else(*default_port),
            pu_authority_username,
            pu_authority_password,
        );
```

and move pu_path and pu_query close  to the code block where they are first used